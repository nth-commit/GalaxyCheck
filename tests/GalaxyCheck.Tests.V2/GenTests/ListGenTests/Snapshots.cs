using GalaxyCheck;
using GalaxyCheck.Gens;
using Snapshooter;
using Snapshooter.Xunit;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using Xunit;

namespace Tests.V2.GenTests.ListGenTests
{
    public class Snapshots
    {
        [Fact]
        public void Values()
        {
            var seeds = Enumerable.Range(0, 5);

            var gens = new Dictionary<string, IListGen<int>>
            {
                { "Between_Lengths_1_And_4_And_Elements_0_And_10", Gen.Int32().Between(0, 10).ListOf().BetweenLengths(1, 4) },
                { "Between_Lengths_1_And_8_And_Elements_0_And_100", Gen.Int32().Between(0, 100).ListOf().BetweenLengths(1, 8) },
                { "Between_Lengths_1_And_16_And_Elements_0_And_10000", Gen.Int32().Between(0, 10000).ListOf().BetweenLengths(1, 16) },
                { "Between_Defaults", Gen.Int32().ListOf() },
            };

            var biases = new[] { Gen.Bias.None, Gen.Bias.WithSize };

            foreach (var seed in seeds)
            foreach (var (betweenDescription, gen) in gens)
            foreach (var bias in biases)
            {
                var gen0 = gen.WithLengthBias(bias);

                var sample = gen.Sample(seed: seed);

                var nameExtension = string.Join("_", new[]
                {
                    betweenDescription,
                    $"Bias_{bias}",
                    $"Seed_{seed}"
                });

                Snapshot.Match(sample, SnapshotNameExtension.Create(nameExtension));
            }
        }

        [Fact]
        public void ExampleSpaces()
        {
            var seeds = Enumerable.Range(0, 5);

            var gens = new Dictionary<string, IListGen<int>>
            {
                { "Between_Lengths_1_And_4_And_Elements_0_And_10", Gen.Int32().Between(0, 10).ListOf().BetweenLengths(1, 4) },
                { "Between_Lengths_1_And_8_And_Elements_0_And_100", Gen.Int32().Between(0, 100).ListOf().BetweenLengths(1, 8) },
            };

            var biases = new[] { Gen.Bias.None, Gen.Bias.WithSize };

            foreach (var seed in seeds)
            foreach (var (betweenDescription, gen) in gens)
            foreach (var bias in biases)
            {
                var gen0 = gen.WithLengthBias(bias);

                var sample = gen.Advanced
                    .SampleOneExampleSpace(seed: seed, size: 75)
                    .Take(500)
                    .Render(x => JsonSerializer.Serialize(x));

                var nameExtension = string.Join("_", new[]
                {
                    betweenDescription,
                    $"Bias_{bias}",
                    $"Seed_{seed}"
                });

                Snapshot.Match(sample, SnapshotNameExtension.Create(nameExtension));
            }
        }
    }
}
