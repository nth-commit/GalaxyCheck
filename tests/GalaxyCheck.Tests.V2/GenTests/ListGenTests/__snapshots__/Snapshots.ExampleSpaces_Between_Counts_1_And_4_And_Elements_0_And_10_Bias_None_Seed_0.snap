[0,3] (Id = 1372007287, Distance = 63.3333...)
├> [3,0] (Id = -149127011, Distance = 63.3333...)
|  ├> [3] (Id = -1521134292, Distance = 30.0)
|  |  ├> [0] (Id = -1521134295, Distance = 0)
|  |  └> [2] (Id = -1521134293, Distance = 20.0)
|  |     └> [1] (Id = -1521134294, Distance = 10.0)
|  ├> [0] (Id = -1521134295, Distance = 0)
|  ├> [0,0] (Id = 119308578, Distance = 33.3333...)
|  └> [2,0] (Id = 1372007284, Distance = 53.3333...)
|     ├> [2] (Id = -1521134293, Distance = 20.0)
|     |  └> [1] (Id = -1521134294, Distance = 10.0)
|     └> [1,0] (Id = -1401825717, Distance = 43.3333...)
|        └> [1] (Id = -1521134294, Distance = 10.0)
├> [0] (Id = -1521134295, Distance = 0)
├> [3] (Id = -1521134292, Distance = 30.0)
|  └> [2] (Id = -1521134293, Distance = 20.0)
|     └> [1] (Id = -1521134294, Distance = 10.0)
├> [0,0] (Id = 119308578, Distance = 33.3333...)
└> [0,2] (Id = 119308580, Distance = 53.3333...)
   ├> [2,0] (Id = 1372007284, Distance = 53.3333...)
   |  ├> [2] (Id = -1521134293, Distance = 20.0)
   |  |  └> [1] (Id = -1521134294, Distance = 10.0)
   |  └> [1,0] (Id = -1401825717, Distance = 43.3333...)
   |     └> [1] (Id = -1521134294, Distance = 10.0)
   ├> [2] (Id = -1521134293, Distance = 20.0)
   |  └> [1] (Id = -1521134294, Distance = 10.0)
   └> [0,1] (Id = 119308579, Distance = 43.3333...)
      ├> [1,0] (Id = -1401825717, Distance = 43.3333...)
      |  └> [1] (Id = -1521134294, Distance = 10.0)
      └> [1] (Id = -1521134294, Distance = 10.0)
